### DISCUSSION TITLE
Parsing a date

### DISCUSSION THREAD

**Discussion by blocka** [colinhacks/zod#879] [Q&A] [answered]
Very surprised that I could not find anything on this.

If I have a date schema `const schema = z.date()`  and I pass it a date string `schema.parse('2022-01-12T00:00:00.000Z')`
I would expect to get back a JS Date object. Instead I get an exception that I've passed a string instead of a date.

Am I supposed to use a transform here?

**Comment by JacobWeisenburger** [comment]
There are a few ways to handle what you are trying to do:

```ts
const dateSchema = z.preprocess( arg => typeof arg == 'string' ? new Date( arg ) : undefined, z.date() )
dateSchema.safeParse( '2022-01-12T00:00:00.000Z' ) // success: true
dateSchema.safeParse( new Date( '2022-01-12T00:00:00.000Z' ) ) // success: false
```

```ts
const stringToDate = z.string().transform( arg => new Date( arg ) )
stringToDate.safeParse( '2022-01-12T00:00:00.000Z' ) // success: true
dateSchema.safeParse( new Date( '2022-01-12T00:00:00.000Z' ) ) // success: false
```

```ts
const stringOrDateToDate = z.string().or( z.date() ).transform( arg => new Date( arg ) )
stringOrDateToDate.safeParse( '2022-01-12T00:00:00.000Z' ) // success: true
stringOrDateToDate.safeParse( new Date( '2022-01-12T00:00:00.000Z' ) ) // success: true
```

Hope this helps. Let me know if you have any questions.

I also added a pull request to add documentation about z.date. Sorry for the confusion cause by not having any documentation.
https://github.com/colinhacks/zod/pull/881

---
2022-02-11 update:
a dates section has now been added to the README
https://github.com/colinhacks/zod#dates

**Comment by Zegnat** [comment]
I ran into this today and thought I would add my solution to this discussion in case others come by:

```js
const stringToValidDate = z.string().transform((dateString, ctx) => {
    const date = new Date(dateString);
    if (!z.date().safeParse(date).success) {
        ctx.addIssue({
            code: z.ZodIssueCode.invalid_date,
        })
    }
    return date;
})
```

I did not want to use `preprocess` as above, as it seemed to effect the `z.input` type function. (Or maybe I was just doing it wrong.) And the two `transform` alternatives given by @JacobWeisenburger could actually lead to invalid dates, meaning even after Zod had parsed the input string it would not be safe to run certain methods.

This transform covers all the cases I need covered. Zod expects a string value in (as I am getting ISO strings from an external API) and will output a Date object only if the string is parsable into a valid date.

**Reply by baloodevil** [reply]
Is there a way to have `stringToValidDate` be of type `ZodDate` (from `z.date()`) instead of `ZodString` (from `z.string()`)?

**Reply by baloodevil** [reply]
Wow!  Zod is an incredible library.

Use `coerce` to do exactly what I was looking for...

```typescript
const stringToValidDate = z.coerce.date().transform((dateString, ctx) => {
    const date = new Date(dateString);
    if (!z.date().safeParse(date).success) {
        ctx.addIssue({
            code: z.ZodIssueCode.invalid_date,
        })
    }
    return date;
})
```

**Comment by JacobWeisenburger** [comment, marked as answer]
Since v3.20, this got way easier.

Let me know if this works for you?

Code updated on 2023-02-25. Thanks to @pilotmoon, for the idea.
```ts
const stringToDate = z.string().pipe( z.coerce.date() )
console.log( stringToDate.safeParse( '2023-01-10' ).success ) // true
console.log( stringToDate.safeParse( '2023-13-10' ).success ) // false
console.log( stringToDate.safeParse( null ).success ) // false
```

---

the `dates` section has been updated in the README
https://github.com/colinhacks/zod#dates

**Reply by pke** [reply]
@JacobWeisenburger any idea how to support offsets in coerced date types?
With `z.datetime({ offset: true })` is allowed. Any reson why offsets are disabled by default in parsing ISO dates?

**Reply by JacobWeisenburger** [reply]
> @JacobWeisenburger any idea how to support offsets in coerced date types?

Is this what you are looking for?

```ts
const stringToDate = z.string().datetime( { offset: true } ).pipe( z.coerce.date() )
console.log( stringToDate.safeParse( '2023-04-05T00:00:00Z' ).success ) // true
console.log( stringToDate.safeParse( '2023-04-05T00:00:00+02:00' ).success ) // true
console.log( stringToDate.safeParse( '2023-04-05T00:00:00.123+02:00' ).success ) // true
console.log( stringToDate.safeParse( '2023-04-05T00:00:00.123+0200' ).success ) // true
console.log( stringToDate.safeParse( '2023-04-05' ).success ) // false
console.log( stringToDate.safeParse( null ).success ) // false
```

---


> Any reson why offsets are disabled by default in parsing ISO dates?

No, idea. Not sure who made that decision.


---

If you found my answer satisfactory, please consider supporting me. Even a small amount is greatly appreciated. Thanks friend! 🙏
https://github.com/sponsors/JacobWeisenburger

**Reply by pke** [reply]
That new piping interface is worth something ;) Thanks a lot @JacobWeisenburger! 

I have to dig deeper into how it works.
So the `coerce.data()` returns a class whose constructor is fed with the result of the datetime parse?

**Reply by McGern** [reply]
Hi,

Is there any advice on how to make dates like "2000-04-31" fail? new Date() will return the first day of the next month, so the date will not fail. In the past I have created the date then parsed it into the same format and checked the string values e.g. 
"2000-04-31" => May 1st 2000 => formats to "2000-05-01" which is not the same as the original input so can be tested and then fails.

**Comment by pilotmoon** [comment]
According to the readme as it stands today, you can now do:

```javascript
const dateValidatorAcceptsString = z.coerce.date();
```

I don't see any information about `pipe()` in the docs.

**Reply by JacobWeisenburger** [reply]
`coerce` allows the schema to accept any input. I wrote my code the way I did so it only accepts strings as input.

**Reply by JacobWeisenburger** [reply]
Although, I suppose you could to it this way:

```ts
const stringToDate = z.string().pipe( z.coerce.date() )
console.log( stringToDate.safeParse( '2023-01-10' ).success ) // true
console.log( stringToDate.safeParse( '2023-13-10' ).success ) // false
console.log( stringToDate.safeParse( null ).success ) // false
```

**Comment by pcostanz** [comment]
@JacobWeisenburger this seems to be a good solution for my usecase of validating that dates are date-like. I'm trying to generate types from Zod objects that make use of this via `z.infer<typeof ZObject>` where some of the keys use the `stringToDate` the issue I'm having is that in the inferred types they're coming through as `Date` types when the value is actually a string that looks like `2024-01-17`. Would you expect these to be typed as Date or string after inference? I can obviously handle it by converting those to Dates, but it kind of seems like it might defeat the purpose of the piped coerce. Thanks!

Eg:

```
const stringToDate = z.string().pipe(z.coerce.date());

export const ZSeason = z.object({
  SeasonId: z.string(),
  Title: z.string(),
  StartDate: stringToDate,
  EndDate: stringToDate,
});

export type TSeason = z.infer<typeof ZSeason>;

// which then produces
type TSeason = {
    SeasonId?: string;
    Title?: string;
    StartDate?: Date;
    EndDate?: Date;
}
```
